common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/SilentWebViewAuthorizationFragment.kt

```
@@ -0,0 +1,92 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.providers.oauth2
+
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.lifecycle.lifecycleScope
+import com.microsoft.identity.common.adal.internal.AuthenticationConstants.AuthorizationIntentKey.WEB_VIEW_SILENT_AUTHORIZATION_FLOW_TIMEOUT
+import com.microsoft.identity.common.java.providers.RawAuthorizationResult
+import com.microsoft.identity.common.logging.Logger
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+/**
+ * Fragment for handling silent web view authorization flows. It hides all UI and makes it transparent.
+ * This is to support scenario where the app needs to perform authorization without expecting user interaction,
+ * and notifying user. It is equivalent to the silent flow. Such flow is non-ideal but may be required by protocol.
+ * Current requirement is for SWAG (Sign-in-with Apple/Google) flow for Copilot, to allow them to silently migrate
+ * users from their stack to MSA.
+ */
+class SilentWebViewAuthorizationFragment : WebViewAuthorizationFragment() {
+
+    companion object {
+        private const val TAG = "SilentWebViewAuthorizationFragment"
+        private const val DEFAULT_WEB_VIEW_SILENT_AUTHORIZATION_FLOW_TIME_OUT = 5000L
+        private const val MIN_WEB_VIEW_SILENT_AUTHORIZATION_FLOW_TIMEOUT = 1000L
```

Supposing something goes wrong on the MSA side and the user can't be silently signed in, will MSA immediately send back an error, or is there any chance in which we just time out?

---

yes, MSA flow should return an error, which would be part of RawAuthorizationResult. In case it is not, then timeout helps break the flow. Timeout is exception case indicating a bug.